File: frontend/apps/ios/Pantopus/Features/TransactionReviews/ReceivedReviewsViewModel.swift
//
//  ReceivedReviewsViewModel.swift
//  Pantopus
//
//  BLOCK 2D — Marketplace transaction reviews. Drives the "received reviews"
//  section on a profile from `GET /api/transaction-reviews/user/:userId`
//  (route `backend/routes/transactionReviews.js:168`). The endpoint returns
//  the overall average + total; the star histogram and the per-criterion
//  breakdown (communication / accuracy / punctuality) are aggregated here
//  from the individual rows.
//

import Foundation
import Observation

/// Average + count for one optional sub-rating criterion.
public struct CriterionAverage: Sendable, Equatable {
    public let average: Double
    public let count: Int
}

/// One projected review row.
public struct ReceivedReviewRow: Sendable, Equatable, Identifiable {
    public let id: String
    public let reviewerName: String
    public let initials: String
    public let avatarURL: URL?
    public let rating: Int
    public let comment: String?
    public let contextLabel: String
    public let roleLabel: String?
    public let timestamp: String
}

/// Aggregated summary backing the loaded state.
public struct ReceivedReviewsSummary: Sendable, Equatable {
    public let average: Double
    public let total: Int
    /// Five fractions in `0...1`, ordered 5★→1★, for `RatingDistribution`.
    public let distribution: [Double]
    public let communication: CriterionAverage?
    public let accuracy: CriterionAverage?
    public let punctuality: CriterionAverage?
    public let rows: [ReceivedReviewRow]
}

@Observable
@MainActor
public final class ReceivedReviewsViewModel {
    public enum State: Sendable {
        case loading
        case empty
        case loaded(ReceivedReviewsSummary)
        case error(message: String)
    }

    public private(set) var state: State = .loading

    private let userId: String
    private let api: APIClient
    private let now: @Sendable () -> Date
    private var loadedOnce = false

    init(
        userId: String,
        api: APIClient = .shared,
        now: @escaping @Sendable () -> Date = { Date() }
    ) {
        self.userId = userId
        self.api = api
        self.now = now
    }

    public func load() async {
        if !loadedOnce { state = .loading }
        await fetch()
    }

    public func refresh() async {
        await fetch()
    }

    private func fetch() async {
        do {
            let response: TransactionReviewsResponse = try await api.request(
                TransactionReviewsEndpoints.userReviews(userId: userId)
            )
            loadedOnce = true
            if response.reviews.isEmpty {
                state = .empty
            } else {
                state = .loaded(Self.summarize(response, now: now()))
            }
        } catch {
            if !loadedOnce {
                state = .error(
                    message: (error as? APIError)?.errorDescription ?? "Couldn't load reviews."
                )
            }
        }
    }

    // MARK: - Pure aggregation (test surface)

    public static func summarize(
        _ response: TransactionReviewsResponse,
        now: Date
    ) -> ReceivedReviewsSummary {
        let reviews = response.reviews
        let total = reviews.count

        var buckets = [Int](repeating: 0, count: 5) // index 0 == 5★ … index 4 == 1★
        for review in reviews {
            let clamped = min(max(review.rating, 1), 5)
            buckets[5 - clamped] += 1
        }
        let distribution = buckets.map { total > 0 ? Double($0) / Double(total) : 0 }

        return ReceivedReviewsSummary(
            average: response.averageRating,
            total: response.total > 0 ? response.total : total,
            distribution: distribution,
            communication: criterionAverage(reviews.compactMap(\.communicationRating)),
            accuracy: criterionAverage(reviews.compactMap(\.accuracyRating)),
            punctuality: criterionAverage(reviews.compactMap(\.punctualityRating)),
            rows: reviews.map { row(for: $0, now: now) }
        )
    }

    static func criterionAverage(_ values: [Int]) -> CriterionAverage? {
        guard !values.isEmpty else { return nil }
        let sum = values.reduce(0, +)
        let average = (Double(sum) / Double(values.count) * 100).rounded() / 100
        return CriterionAverage(average: average, count: values.count)
    }

    static func row(for dto: TransactionReviewDTO, now: Date) -> ReceivedReviewRow {
        let name = displayName(for: dto.reviewer)
        let context = TransactionReviewContext.fromRaw(dto.context)
        return ReceivedReviewRow(
            id: dto.id,
            reviewerName: name,
            initials: initials(from: name),
            avatarURL: dto.reviewer?.profilePictureUrl.flatMap(URL.init(string:)),
            rating: min(max(dto.rating, 0), 5),
            comment: dto.comment?.isEmpty == false ? dto.comment : nil,
            contextLabel: context?.shortLabel ?? "Transaction",
            roleLabel: roleLabel(isBuyer: dto.isBuyer),
            timestamp: relativeTime(dto.createdAt, now: now)
        )
    }

    static func displayName(for reviewer: TransactionReviewerDTO?) -> String {
        if let first = reviewer?.firstName, !first.isEmpty {
            if let last = reviewer?.lastName, !last.isEmpty {
                return "\(first) \(last)"
            }
            return first
        }
        if let username = reviewer?.username, !username.isEmpty { return username }
        return "Neighbor"
    }

    static func initials(from name: String) -> String {
        let parts = name.split(separator: " ").prefix(2)
        let letters = parts.compactMap(\.first).map(String.init)
        let joined = letters.joined().uppercased()
        return joined.isEmpty ? "?" : joined
    }

    static func roleLabel(isBuyer: Bool?) -> String? {
        switch isBuyer {
        case true: "From buyer"
        case false: "From seller"
        default: nil
        }
    }

    static func relativeTime(_ raw: String?, now: Date) -> String {
        guard let date = parseDate(raw) else { return "" }
        let interval = now.timeIntervalSince(date)
        if interval < 60 { return "just now" }
        if interval < 3600 { return "\(Int(interval / 60))m ago" }
        if interval < 86400 { return "\(Int(interval / 3600))h ago" }
        let days = Int(interval / 86400)
        if days == 1 { return "yesterday" }
        if days < 7 { return "\(days)d ago" }
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "MMM d"
        return formatter.string(from: date)
    }

    private static let iso8601: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    private static let iso8601NoFraction: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    static func parseDate(_ raw: String?) -> Date? {
        guard let raw, !raw.isEmpty else { return nil }
        return iso8601.date(from: raw) ?? iso8601NoFraction.date(from: raw)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/transaction_reviews/ReceivedReviewsViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber")

package app.pantopus.android.ui.screens.transaction_reviews

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.transaction_reviews.TransactionReviewDto
import app.pantopus.android.data.api.models.transaction_reviews.TransactionReviewerDto
import app.pantopus.android.data.api.models.transaction_reviews.TransactionReviewsResponse
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.api.net.displayMessage
import app.pantopus.android.data.transaction_reviews.TransactionReviewsRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale
import javax.inject.Inject
import kotlin.math.roundToInt

/** Average + count for one optional sub-rating criterion. */
data class CriterionAverage(
    val average: Double,
    val count: Int,
)

/** One projected received-review row. */
data class ReceivedReviewRow(
    val id: String,
    val reviewerName: String,
    val initials: String,
    val avatarUrl: String?,
    val rating: Int,
    val comment: String?,
    val contextLabel: String,
    val roleLabel: String?,
    val timestamp: String,
)

/** Aggregated summary backing the loaded state. */
data class ReceivedReviewsSummary(
    val average: Double,
    val total: Int,
    /** Five fractions in 0..1, ordered 5★→1★, for [RatingDistribution]. */
    val distribution: List<Float>,
    val communication: CriterionAverage?,
    val accuracy: CriterionAverage?,
    val punctuality: CriterionAverage?,
    val rows: List<ReceivedReviewRow>,
)

/** Four render states for the received-reviews section. */
sealed interface ReceivedReviewsUiState {
    data object Loading : ReceivedReviewsUiState

    data object Empty : ReceivedReviewsUiState

    data class Loaded(val summary: ReceivedReviewsSummary) : ReceivedReviewsUiState

    data class Error(val message: String) : ReceivedReviewsUiState
}

/**
 * BLOCK 2D — drives the "received reviews" section from
 * `GET /api/transaction-reviews/user/:userId`. The user id is supplied by
 * the embedding composable via [load] (the section isn't a nav destination,
 * so it can't read a SavedStateHandle arg).
 */
@HiltViewModel
class ReceivedReviewsViewModel
    @Inject
    constructor(
        private val repo: TransactionReviewsRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<ReceivedReviewsUiState>(ReceivedReviewsUiState.Loading)
        val state: StateFlow<ReceivedReviewsUiState> = _state.asStateFlow()

        private var userId: String? = null
        private var loadedOnce = false

        fun load(userId: String) {
            if (this.userId == userId && loadedOnce) return
            this.userId = userId
            reload()
        }

        fun refresh() = reload()

        private fun reload() {
            val uid = userId ?: return
            if (!loadedOnce) _state.value = ReceivedReviewsUiState.Loading
            viewModelScope.launch {
                when (val result = repo.userReviews(uid)) {
                    is NetworkResult.Success -> {
                        loadedOnce = true
                        val data = result.data
                        _state.value =
                            if (data.reviews.isEmpty()) {
                                ReceivedReviewsUiState.Empty
                            } else {
                                ReceivedReviewsUiState.Loaded(summarize(data, Instant.now()))
                            }
                    }
                    is NetworkResult.Failure -> {
                        if (!loadedOnce) {
                            _state.value = ReceivedReviewsUiState.Error(result.error.displayMessage("Couldn't load reviews."))
                        }
                    }
                }
            }
        }

        companion object {
            /** Pure aggregation — public so tests assert it without the VM. */
            fun summarize(
                response: TransactionReviewsResponse,
                now: Instant,
            ): ReceivedReviewsSummary {
                val reviews = response.reviews
                val total = reviews.size
                val buckets = IntArray(5) // index 0 == 5★ … index 4 == 1★
                for (review in reviews) {
                    val clamped = review.rating.coerceIn(1, 5)
                    buckets[5 - clamped] += 1
                }
                val distribution = buckets.map { if (total > 0) it.toFloat() / total else 0f }
                return ReceivedReviewsSummary(
                    average = response.averageRating,
                    total = if (response.total > 0) response.total else total,
                    distribution = distribution,
                    communication = criterionAverage(reviews.mapNotNull { it.communicationRating }),
                    accuracy = criterionAverage(reviews.mapNotNull { it.accuracyRating }),
                    punctuality = criterionAverage(reviews.mapNotNull { it.punctualityRating }),
                    rows = reviews.map { row(it, now) },
                )
            }

            fun criterionAverage(values: List<Int>): CriterionAverage? {
                if (values.isEmpty()) return null
                val average = (values.sum().toDouble() / values.size * 100).roundToInt() / 100.0
                return CriterionAverage(average = average, count = values.size)
            }

            fun row(
                dto: TransactionReviewDto,
                now: Instant,
            ): ReceivedReviewRow {
                val name = displayName(dto.reviewer)
                val context = TransactionReviewContext.fromRaw(dto.context)
                return ReceivedReviewRow(
                    id = dto.id,
                    reviewerName = name,
                    initials = initials(name),
                    avatarUrl = dto.reviewer?.profilePictureUrl,
                    rating = dto.rating.coerceIn(0, 5),
                    comment = dto.comment?.takeIf { it.isNotEmpty() },
                    contextLabel = context?.shortLabel ?: "Transaction",
                    roleLabel = roleLabel(dto.isBuyer),
                    timestamp = relativeTime(dto.createdAt, now),
                )
            }

            fun displayName(reviewer: TransactionReviewerDto?): String {
                val first = reviewer?.firstName?.takeIf { it.isNotEmpty() }
                if (first != null) {
                    val last = reviewer.lastName?.takeIf { it.isNotEmpty() }
                    return if (last != null) "$first $last" else first
                }
                val username = reviewer?.username?.takeIf { it.isNotEmpty() }
                if (username != null) return username
                return "Neighbor"
            }

            fun initials(name: String): String {
                val letters =
                    name
                        .split(" ")
                        .filter { it.isNotBlank() }
                        .take(2)
                        .mapNotNull { it.firstOrNull()?.toString() }
                val joined = letters.joinToString("").uppercase(Locale.ROOT)
                return joined.ifEmpty { "?" }
            }

            fun roleLabel(isBuyer: Boolean?): String? =
                when (isBuyer) {
                    true -> "From buyer"
                    false -> "From seller"
                    null -> null
                }

            fun relativeTime(
                raw: String?,
                now: Instant,
            ): String {
                val date = parseInstant(raw) ?: return ""
                val seconds = ChronoUnit.SECONDS.between(date, now)
                return when {
                    seconds < 60 -> "just now"
                    seconds < 3600 -> "${seconds / 60}m ago"
                    seconds < 86_400 -> "${seconds / 3600}h ago"
                    seconds < 2 * 86_400 -> "yesterday"
                    seconds < 7 * 86_400 -> "${seconds / 86_400}d ago"
                    else -> monthDay(date)
                }
            }

            private fun monthDay(instant: Instant): String =
                runCatching {
                    DateTimeFormatter
                        .ofPattern("MMM d", Locale.US)
                        .withZone(java.time.ZoneId.systemDefault())
                        .format(instant)
                }.getOrDefault("")

            private fun parseInstant(raw: String?): Instant? {
                if (raw.isNullOrEmpty()) return null
                return runCatching { Instant.parse(raw) }.getOrNull()
            }
        }
    }
